In [ ]:
!pip install python-dotenv pandas numpy openpyxl sqlalchemy

In [1]:
import os
from dotenv import load_dotenv

load_dotenv()

DB_HOST = os.getenv("SANDHYA_ERP_DB_HOST")
DB_PORT = os.getenv("SANDHYA_ERP_DB_PORT")

DB_USERNAME = os.getenv("SANDHYA_ERP_DB_USERNAME")
DB_PASSWORD = os.getenv("SANDHYA_ERP_DB_PASSWORD")
DB_NAME = os.getenv("SANDHYA_ERP_DB_NAME")

sandhya_erp_db_url = (
    f"mysql+pymysql://{DB_USERNAME}:{DB_PASSWORD}@{DB_HOST}:{DB_PORT}/{DB_NAME}"
)

In [2]:
import pandas as pd
import sqlalchemy

# show all rows and columns
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

In [3]:
engine = sqlalchemy.create_engine(sandhya_erp_db_url)

In [4]:
supplier_df = pd.read_sql("SELECT * FROM erpx_dev.suppliers", engine)
supplier_df.head()

supplier_id code                      name supplier_category gst_number  \
0            1   A1               Karuna Raju             Agent       None   
1            2   A2    Sri Lakshmi Aqua Farms             Agent       None   
2            3   A3  Sri Venkateswara Traders             Agent       None   
3            4   A4      Varalakshmi Agencies             Agent       None   
4            7   A5    Sri Lakshmi Aqua Farms             Agent       None   

   broker_id contact_person street apartment  city state district postal_code  \
0          0      Gopi Raju   None      None  None    ap   Ongole      123456   
1          0        K Vijay   None      None  None    ap   bhVEM       123456   
2          0            SVT   None      None  None    ap   bhVEM       123456   
3          0         Rakesh   None      None  None    ap       KK      123456   
4          0         Rakesh   None      None  None    AP       KK      123456   

   commission bank_acc_no bank_ifsc bank_name bank_branch contact_number  \
0         2.0        None      None      None        None           None   
1         2.0        None      None      None        None           None   
2         2.0        None      None      None        None           None   
3         1.0        None      None      None        None           None   
4         1.0        None      None      None        None           None   

  email_id  is_active                 created_at created_user  \
0     None          1 2023-01-20 17:18:17.358860         None   
1     None          1 2023-01-21 15:59:41.769207         None   
2     None          1 2023-01-21 16:19:28.500526         None   
3     None          1 2023-01-23 14:39:43.386196         None   
4     None          0 2023-01-23 17:33:11.528853         None   

                  updated_at updated_user  version_flag payment_period  
0 2023-01-23 15:24:21.920682         None             1           None  
1 2023-01-23 15:24:27.197321         None             1           None  
2 2023-01-23 15:24:30.204286         None             1           None  
3 2023-01-23 15:24:34.793301         None             1           None  
4 2023-01-24 19:23:35.251509         None             2           None

In [5]:
supplier_df.describe().T

count                           mean  \
supplier_id   572.0                     288.486014   
broker_id     572.0                       0.215035   
commission    567.0                     920.155203   
is_active     572.0                       0.977273   
created_at      572  2023-08-27 04:53:19.229318400   
updated_at      572  2023-09-01 05:44:39.985777152   
version_flag  572.0                       1.083916   

                                     min                            25%  \
supplier_id                          1.0                         145.75   
broker_id                            0.0                            0.0   
commission                           1.0                            1.0   
is_active                            0.0                            1.0   
created_at    2023-01-20 17:18:17.358860  2023-02-25 11:28:33.746909440   
updated_at    2023-01-23 15:24:21.920682  2023-02-25 15:49:37.842742528   
version_flag                         1.0                            1.0   

                                        50%                            75%  \
supplier_id                           288.5                         431.25   
broker_id                               0.0                            0.0   
commission                              1.0                            1.0   
is_active                               1.0                            1.0   
created_at    2023-06-01 14:03:45.611426048  2023-11-17 12:35:04.927335936   
updated_at    2023-06-03 04:25:53.728720384  2023-11-28 12:55:13.254458880   
version_flag                            1.0                            1.0   

                                     max           std  
supplier_id                        574.0     165.29061  
broker_id                          123.0      5.142888  
commission                      521157.0  21886.494395  
is_active                            1.0      0.149163  
created_at    2025-08-12 12:22:13.410409           NaN  
updated_at    2025-08-12 12:22:13.410409           NaN  
version_flag                         4.0      0.307444

No significant numerical data in erpx_dev.suppliers table

In [6]:
# supplier with max commission

supplier_df[supplier_df["commission"] == supplier_df["commission"].max()]

supplier_id  code             name supplier_category gst_number  \
489          492  A481  SUMUKHA TRADERS             Agent       None   

     broker_id contact_person street apartment  city state district  \
489          0         PARASA   None      None  None    AP  KRISHNA   

    postal_code  commission bank_acc_no bank_ifsc bank_name bank_branch  \
489        None    521157.0        None      None      None        None   

    contact_number email_id  is_active                 created_at  \
489           None     None          1 2024-07-08 17:07:10.619320   

    created_user                 updated_at updated_user  version_flag  \
489         None 2024-07-08 17:07:10.619320         None             1   

    payment_period  
489           None

In [7]:
supplier_df.nunique()

supplier_id          572
code                 572
name                 556
supplier_category      3
gst_number             1
broker_id              2
contact_person        11
street                 1
apartment              0
city                   1
state                  5
district               7
postal_code            7
commission             3
bank_acc_no            0
bank_ifsc              0
bank_name              0
bank_branch            0
contact_number         1
email_id               0
is_active              2
created_at           572
created_user           0
updated_at           572
updated_user           0
version_flag           4
payment_period         0
dtype: int64

In [8]:
supplier_df["supplier_category"].value_counts()

supplier_category
Agent     561
Farmer     10
Broker      1
Name: count, dtype: int64

Among 572 suppliers, 561 are agents, which is almost all of the suppliers. 

In [9]:
supplier_df[supplier_df["supplier_category"] == "Broker"]

supplier_id code   name supplier_category gst_number  broker_id  \
120          123   B1  asdff            Broker       None          0   

    contact_person street apartment  city state district postal_code  \
120           None   None      None  None  None     None        None   

     commission bank_acc_no bank_ifsc bank_name bank_branch contact_number  \
120         NaN        None      None      None        None           None   

    email_id  is_active                 created_at created_user  \
120     None          1 2023-02-21 16:24:21.113710         None   

                    updated_at updated_user  version_flag payment_period  
120 2023-02-21 16:24:21.113710         None             1           None

In [10]:
supplier_df[supplier_df["supplier_category"] == "Farmer"]

supplier_id  code                                           name  \
5              8    F1                                         K Sriu   
6              9    F2                                 srinivasarao k   
21            24    F3                                          rahul   
83            86    F4                                         madhav   
234          237  A230             SANDHYA AQUA FARMS KRITHIVENNU OWN   
359          362    F5                                    V DURGA RAO   
414          417    F6  SANDHYA AQUA EXPORTS PVT LTD FARM BEERAMGUNTA   
437          440    F7                         SANDHYA AQUA SRI FARMS   
487          490    F8                               GUDURU OWN PONDS   
561          564    F9                                              x   

    supplier_category   gst_number  broker_id contact_person street apartment  \
5              Farmer         None          0         Rakesh   None      None   
6              Farmer         None          0        sravani   None      None   
21             Farmer  GST33654789          0           None   None      None   
83             Farmer         None        123           None   None      None   
234            Farmer         None          0           None   None      None   
359            Farmer         None          0        SRAVANI   None      None   
414            Farmer         None          0        KAVITHA   None      None   
437            Farmer         None          0        KAVITHA   None      None   
487            Farmer         None          0         PARASA   None      None   
561            Farmer         None          0           None   None      None   

     city state  district postal_code  commission bank_acc_no bank_ifsc  \
5    None    AP        KK      123456         1.0        None      None   
6    None    AP   Krishna      521157         NaN        None      None   
21   None  None      None        None         NaN        None      None   
83   None  None      None        None         NaN        None      None   
234  None  None      None        None         1.0        None      None   
359  None  None   KRISHNA      521157         1.0        None      None   
414  None    AP  KRISHANA      521157         1.0        None      None   
437  None    AP   KRISHNA      521157         1.0        None      None   
487  None    AP   KRISHNA      521157         1.0        None      None   
561  None  None      None        None         NaN        None      None   

    bank_name bank_branch contact_number email_id  is_active  \
5        None        None           None     None          1   
6        None        None           None     None          1   
21       None        None           None     None          1   
83       None        None           None     None          1   
234      None        None           None     None          1   
359      None        None           None     None          1   
414      None        None           None     None          1   
437      None        None           None     None          1   
487      None        None           None     None          1   
561      None        None           None     None          1   

                    created_at created_user                 updated_at  \
5   2023-01-24 11:08:26.341499         None 2023-01-24 11:08:26.341499   
6   2023-01-24 11:36:09.012417         None 2023-01-24 19:29:02.666640   
21  2023-01-24 19:38:08.526412         None 2023-01-24 19:38:08.526412   
83  2023-02-08 18:23:33.541614         None 2023-02-21 16:24:21.000000   
234 2023-04-21 10:41:13.773245         None 2023-08-29 17:01:45.000000   
359 2023-08-01 10:20:03.658414         None 2023-08-01 10:20:03.658414   
414 2023-10-21 14:12:00.607933         None 2023-10-21 14:12:00.607933   
437 2023-12-12 11:09:54.815987         None 2023-12-12 11:09:54.815987   
487 2024-06-27 12:43:06.874291         None 2024-06-27 12:43:06.874291   
561 2025-06-18 13:06:43.879680        

In [11]:
indent_df = pd.read_sql("SELECT * FROM erpx_dev_rm_procurement.indent", engine)
indent_df.head()

is_active                 created_at created_user          updated_at  \
0          1 2023-01-20 18:36:47.331519      unitone 2023-01-23 15:43:48   
1          1 2023-01-20 18:48:41.656502      unitone 2023-01-20 18:49:11   
2          1 2023-01-21 16:02:02.897476   superadmin 2023-01-21 16:09:08   
3          1 2023-01-21 16:21:02.870258   superadmin 2023-01-23 10:30:51   
4          1 2023-01-23 14:42:02.614058      unittwo 2023-01-23 14:48:31   

  updated_user  version_flag  indent_id   indent_code         indent_date  \
0      unitone             6          3  I24/A20/0001 2023-01-20 13:05:15   
1      unitone             3          4  I24/A20/0002 2023-01-20 13:18:12   
2   superadmin             5          5  I24/A21/0003 2023-01-21 10:29:58   
3   superadmin             5          6  I24/A21/0004 2023-01-21 10:49:33   
4      unittwo             6          7  I24/A23/0005 2023-01-23 09:09:50   

  supplier_type commission_type  commission     harvesting_date  \
0         Agent      Percentage           0 2023-01-21 13:05:15   
1         Agent      Percentage           0 2023-01-21 13:18:12   
2         Agent      Percentage           0 2023-01-22 10:29:58   
3         Agent      Percentage           0 2023-01-22 10:49:33   
4         Agent      Percentage           0 2023-01-24 09:09:50   

  harvesting_time  assign_lead_1  assign_lead_2  vehicle_assigned  \
0         Morning           1809            NaN                 0   
1         Morning           1809            NaN                 0   
2         Morning           1809            NaN                 0   
3         Morning           1809            NaN                 0   
4         Morning           1809            NaN                 0   

   grn_completed  plant_id  indent_by  is_vehicle_required  harvest_type  
0              1         1       2178                    0  Re Weighment  
1              1         2       1810                    0  Re Weighment  
2              1         2       2178                    0  Live Harvest  
3              1         1       2178                    0  Re Weighment  
4              1         3       1809                    0  Live Harvest

In [12]:
indent_items_df = pd.read_sql(
    "SELECT * FROM erpx_dev_rm_procurement.indent_items", engine
)
indent_items_df.head()

indent_item_id  broker  farmer  farmer_code  product  expected_count  \
0               2       0       1          NaN        2              97   
1               3       0       1          NaN        2              45   
2               4       0       2          NaN        2              25   
3               5       0       3          NaN        2              55   
4               6       0       4          NaN        2              70   

   expected_qty  expected_price  expected_count2  expected_price2  \
0          1500             237            100.0            225.0   
1          1000             350              NaN              NaN   
2          5000             300              NaN              NaN   
3          3500             275             57.0            270.0   
4          2500             290             75.0            286.0   

  expected_count3 expected_price3  is_active                 created_at  \
0            None            None          1 2023-01-20 18:36:47.337082   
1            None            None          1 2023-01-20 18:48:41.662178   
2            None            None          1 2023-01-21 16:02:02.904148   
3            None            None          1 2023-01-21 16:21:02.877064   
4            None            None          1 2023-01-23 14:42:02.627996   

  created_user          updated_at updated_user  version_flag  indent_id  \
0        Admin 2023-01-20 18:37:04         None             2          3   
1        Admin 2023-01-20 18:48:52         None             2          4   
2        Admin 2023-01-21 16:02:11         None             2          5   
3        Admin 2023-01-21 16:21:33         None             2          6   
4        Admin 2023-01-23 14:42:18         None             2          7   

   harvest_report_id  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0

In [13]:
indent_items_df.describe().T

count                           mean  \
indent_item_id     14639.0                         7321.0   
broker             14639.0                            0.0   
farmer             14639.0                     178.764943   
farmer_code            1.0                            1.0   
product            14639.0                       1.952183   
expected_count     14639.0                      57.208757   
expected_qty       14639.0                    2839.682355   
expected_price     14639.0                     316.733042   
expected_count2       12.0                     152.083333   
expected_price2        8.0                        337.375   
is_active          14639.0                            1.0   
created_at           14639  2023-11-23 18:02:02.982646784   
updated_at           14639  2023-11-24 15:38:29.266195712   
version_flag       14639.0                       1.546554   
indent_id          14639.0                    4113.736457   
harvest_report_id  14639.0                            0.0   

                                          min                            25%  \
indent_item_id                            2.0                         3661.5   
broker                                    0.0                            0.0   
farmer                                    1.0                           39.0   
farmer_code                               1.0                            1.0   
product                                   1.0                            2.0   
expected_count                            1.0                           35.0   
expected_qty                             10.0                         1100.0   
expected_price                            0.0                          265.0   
expected_count2                          34.0                          52.75   
expected_price2                          90.0                         258.75   
is_active                                 1.0                            1.0   
created_at         2023-01-20 18:36:47.337082  2023-05-12 11:44:58.126791168   
updated_at                2023-01-20 18:37:04  2023-05-12 12:04:12.205904896   
version_flag                              1.0                            1.0   
indent_id                                 3.0                         2067.5   
harvest_report_id                         0.0                            0.0   

                                             50%  \
indent_item_id                            7321.0   
broker                                       0.0   
farmer                                     133.0   
farmer_code                                  1.0   
product                                      2.0   
expected_count                              55.0   
expected_qty                              2000.0   
expected_price                             300.0   
expected_count2                            105.0   
expected_price2                            274.0   
is_active                                    1.0   
created_at         2023-08-29 16:03:13.943058944   
updated_at                   2023-08-29 17:22:19   
version_flag                                 1.0   
indent_id                                 4216.0   
harvest_report_id                            0.0   

                                             75%                         max  \
indent_item_id                           10980.5                     14640.0   
broker                                       0.0                         0.0   
farmer                                     284.0                       574.0   
farmer_code                                  1.0                         1.0   
product                                      2.0                         3.0   
expected_count                              75.0                      1800.0   
expected_qty                              3700.0                     36000.0   
expected_price                             350.0                      8000.0   
expected_count2  

In [14]:
indent_items_df.nunique()

indent_item_id       14639
broker                   1
farmer                 551
farmer_code              1
product                  3
expected_count         115
expected_qty           187
expected_price         173
expected_count2         11
expected_price2          7
expected_count3          0
expected_price3          0
is_active                1
created_at           14639
created_user             1
updated_at           11613
updated_user             0
version_flag            10
indent_id             8200
harvest_report_id        1
dtype: int64

In [15]:
grn_df = pd.read_sql("SELECT * FROM erpx_dev_rm_procurement.grn", engine)
grn_df.head()

grn_id            grn_date         grn_number supplier_type  is_active  \
0      10 2023-01-20 18:42:08  G24/KRN/null/0001         Agent          1   
1      11 2023-01-20 18:48:55  G24/KRN/null/0002         Agent          1   
2      12 2023-01-20 18:54:14  G24/KRN/null/0003         Agent          1   
3      13 2023-01-21 16:02:15  G24/123/null/0004         Agent          1   
4      14 2023-01-21 16:21:36  G24/SVT/null/0005         Agent          1   

                  created_at created_user                 updated_at  \
0 2023-01-20 18:43:15.625959      unitone 2023-01-23 12:11:13.000000   
1 2023-01-20 18:49:11.632700      unitone 2023-03-30 14:26:57.151334   
2 2023-01-20 18:54:36.696090      unitone 2023-01-23 15:43:48.000000   
3 2023-01-21 16:06:28.127043   superadmin 2023-01-21 16:09:34.000000   
4 2023-01-21 16:23:02.922709   superadmin 2023-01-23 10:30:51.000000   

  updated_user  version_flag  indent_no  plant_id  sub_plant_id  
0      unitone             4          3         2           6.0  
1         None             1          4         0           NaN  
2        admin             2          3         1           NaN  
3   superadmin             4          5         2           4.0  
4        admin             4          6         1           8.0

In [16]:
grn_items_df = pd.read_sql("SELECT * FROM erpx_dev_rm_procurement.grn_items", engine)
grn_items_df.head()

grn_item_id  grn_id  indent_id         lot_number  sale_order_id  \
0            1      10          3   LOT-U1/24/1/0001            NaN   
1            2      11          4   LOT-U1/24/1/0049            NaN   
2            3      12          3               None            NaN   
3            4      13          5   LOT-U1/24/2/0416            NaN   
4            5      14          6  LOT-KKD/24/1/0002            NaN   

   sale_order_item_id  farmer  farmer_code  quantity  count    price  \
0                 NaN       1          NaN      1000     55  50000.0   
1                 NaN       1          NaN      1000     50      NaN   
2                 NaN       1          NaN      1490     97   5200.0   
3                 NaN       2          NaN      2606     31    290.0   
4                 NaN       3          NaN      3264     57    270.0   

   received_boxes  soft_percentage  boxes  is_active  assign_to_rm  \
0           120.0              2.0    120          1             1   
1             0.0              NaN    150          1             1   
2            48.0              0.0     48          1             0   
3            87.0              0.0     87          1             1   
4           116.0              0.0    116          1             1   

                  created_at created_user          updated_at updated_user  \
0 2023-01-20 18:43:15.631362      unitone 2023-01-23 12:11:18      unitone   
1 2023-01-20 18:49:11.637446      unitone 2023-01-25 12:04:13         None   
2 2023-01-20 18:54:36.702904      unitone 2023-01-23 15:43:48        admin   
3 2023-01-21 16:06:28.135154   superadmin 2023-02-08 15:42:25   superadmin   
4 2023-01-21 16:23:02.927747   superadmin 2023-01-23 12:24:27        admin   

   version_flag  indent_item_id  product  broker  vi_inspection  \
0             3               2        2       0              0   
1             2               3        2       0              0   
2             2               2        2       0              0   
3             4               4        4       0              0   
4             4               5        5       0              0   

   rm_inspection  antibiotic_test actual_quantity actual_count loose_shell  \
0              0                0            None         None           3   
1              0                0            None         None        None   
2              0                0            None         None           0   
3              0                0            None         None           0   
4              0                0            None         None           0   

   fungus_percentage  ice_boxes  empty_boxes  rm_boxes seal_no remarks  \
0                1.0        NaN          NaN       NaN    None    None   
1                NaN        NaN          NaN       NaN    None    None   
2                0.0        NaN          NaN       NaN    None    None   
3                0.0        NaN          NaN       NaN    None    None   
4                0.0        NaN          NaN       NaN    None    None   

  plant_lot_number  ref_no  
0             None     NaN  
1             None     NaN  
2             None     NaN  
3             None     NaN  
4             None     NaN

In [17]:
grn_items_df.describe().T

count                           mean  \
grn_item_id         14564.0                    7427.173647   
grn_id              14564.0                    4269.286116   
indent_id           14564.0                    4111.924746   
sale_order_id       12570.0                    2965.171997   
sale_order_item_id  12570.0                   10932.033413   
farmer              14564.0                     178.641994   
farmer_code             1.0                            1.0   
quantity            14564.0                    2822.714227   
count               14564.0                      60.093244   
price                 269.0                     517.925651   
received_boxes      13782.0                       94.80997   
soft_percentage     12075.0                       0.026335   
boxes               14564.0                      93.192461   
is_active           14564.0                            1.0   
assign_to_rm        14564.0                       0.995468   
created_at            14564  2023-11-24 18:57:33.053271552   
updated_at            14564  2025-02-15 08:55:02.211108352   
version_flag        14564.0                       4.085279   
indent_item_id      14564.0                    7285.468415   
product             14564.0                    6797.585553   
broker              14564.0                            0.0   
vi_inspection       14564.0                       0.000206   
rm_inspection       14564.0                            0.0   
antibiotic_test     14564.0                       0.000069   
fungus_percentage   12075.0                       0.011097   
ice_boxes              66.0                       5.560606   
empty_boxes            66.0                      19.939394   
rm_boxes               66.0                           98.5   
ref_no              12527.0                    1260.550491   

                                           min                            25%  \
grn_item_id                                1.0                        3799.75   
grn_id                                    10.0                        2242.75   
indent_id                                  3.0                        2057.75   
sale_order_id                            220.0                         1927.0   
sale_order_item_id                       358.0                         8397.0   
farmer                                     1.0                           39.0   
farmer_code                                1.0                            1.0   
quantity                                   4.0                         1149.0   
count                                      5.0                           40.0   
price                                    103.0                          278.0   
received_boxes                             0.0                           40.0   
soft_percentage                            0.0                            0.0   
boxes                                      1.0                           39.0   
is_active                                  1.0                            1.0   
assign_to_rm                               0.0                            1.0   
created_at          2023-01-20 18:43:15.631362  2023-05-11 18:18:50.947343360   
updated_at                 2023-01-23 12:11:18  2025-05-09 09:22:22.411624960   
version_flag                               1.0                            3.0   
indent_item_id                             1.0                        3644.75   
product                                    1.0                        2976.75   
broker                                     0.0                            0.0   
vi_inspection                              0.0                            0.0   
rm_inspection                              0.0                            0.0   
antibiotic_test                            0.0                            0.0   
fungus_percentage                          0.0                            0.0   
ice_boxes                                  0.0         

In [18]:
grn_items_df["plant_lot_number"].nunique()

12959

In [19]:
grn_items_df.nunique()

grn_item_id           14564
grn_id                 8169
indent_id              8165
lot_number            13466
sale_order_id           693
sale_order_item_id      801
farmer                  551
farmer_code               1
quantity               5967
count                   175
price                   101
received_boxes          484
soft_percentage          12
boxes                   483
is_active                 1
assign_to_rm              2
created_at            14564
created_user             23
updated_at             2352
updated_user             21
version_flag             17
indent_item_id        14074
product               12072
broker                    1
vi_inspection             2
rm_inspection             1
antibiotic_test           2
actual_quantity           0
actual_count              0
loose_shell              20
fungus_percentage         6
ice_boxes                15
empty_boxes              25
rm_boxes                 52
seal_no                  38
remarks             

In [20]:
query = """
    SELECT

    -- Farmer Information
    ii.farmer,
    s.name AS farmer_name,

    -- Time to GRN
    AVG(TIMESTAMPDIFF(HOUR, i.indent_date, g.grn_date)) AS avg_time_to_grn,

    -- Average Expected (Planned) Values from Indent
    SUM(ii.expected_qty) AS total_expected_qty,                                 -- SUMMING ALL EXPECTED QUANITTY FROM FARMER RATHER TO AVERAGING
    AVG(ii.expected_price) AS avg_expected_price,
    -- AVG(ii.expected_count) AS avg_expected_count, 

    -- Average Actual Received Values from GRN
    SUM(gi.quantity) AS total_actual_quantity,                           -- SUMMING ALL ACTUAL QUANITTY FROM FARMER RATHER TO AVERAGING   
    AVG(gi.price) AS avg_actual_price,
    -- AVG(gi.actual_count) AS avg_actual_count,
    AVG(gi.received_boxes) AS avg_received_boxes,

    -- Key Quality & Performance Metrics
    AVG(ABS(gi.quantity - ii.expected_qty)) AS avg_quantity_variance,
    AVG(gi.price - ii.expected_price) AS avg_price_variance,
    AVG(ABS(gi.count - ii.expected_count)) AS avg_expected_count_variance,
    

    -- Identifiers for Traceability
    GROUP_CONCAT(DISTINCT i.indent_id) AS indent_ids,
    GROUP_CONCAT(DISTINCT ii.indent_item_id) AS indent_item_ids,
    GROUP_CONCAT(DISTINCT g.grn_id) AS grn_ids,
    GROUP_CONCAT(DISTINCT gi.grn_item_id) AS grn_item_ids,
    
    COUNT(DISTINCT i.indent_id) AS indent_count,
    COUNT(DISTINCT g.grn_id) AS grn_count,
    COUNT(DISTINCT gi.indent_item_id) AS indent_item_count,
    COUNT(DISTINCT gi.grn_item_id) AS grn_item_count,

    COUNT(DISTINCT gi.plant_lot_number) AS lots_supplied

FROM
    erpx_dev_rm_procurement.indent_items AS ii
JOIN
    erpx_dev_rm_procurement.indent AS i ON ii.indent_id = i.indent_id
LEFT JOIN
    erpx_dev_rm_procurement.grn AS g ON g.indent_no = i.indent_id
LEFT JOIN 
    erpx_dev_rm_procurement.grn_items AS gi ON gi.grn_id = g.grn_id
LEFT JOIN
    erpx_dev.suppliers AS s ON s.supplier_id = ii.farmer
GROUP BY
    ii.farmer
ORDER BY
    ii.farmer;
"""

test_df = pd.read_sql(query, engine)
test_df.head()

farmer               farmer_name  avg_time_to_grn  total_expected_qty  \
0       1               Karuna Raju         302.8000             21347.0   
1       2    Sri Lakshmi Aqua Farms          33.7000             43700.0   
2       3  Sri Venkateswara Traders           5.0000              3500.0   
3       4      Varalakshmi Agencies         128.7684            367500.0   
4       8                    K Sriu           3.3333             10500.0   

   avg_expected_price  total_actual_quantity  avg_actual_price  \
0            301.4000                23530.0        27600.0000   
1            293.8000                35934.0          278.0000   
2            275.0000                 3264.0          270.0000   
3            319.5526               336530.0          312.5839   
4            286.6667                10072.0          287.5000   

   avg_received_boxes  avg_quantity_variance  avg_price_variance  \
0             84.5556               421.3000          27363.0000   
1            122.0000              1389.8000             -3.6000   
2            116.0000               236.0000             -5.0000   
3             69.5122               852.7263              5.7591   
4             60.8000               343.6667              7.5000   

   avg_expected_count_variance  \
0                       7.3000   
1                       4.5000   
2                       2.0000   
3                       9.6895   
4                       5.3333   

                                          indent_ids  \
0                    3,4,9,56,57,1399,2671,4637,7297   
1                                     5,10,1603,7529   
2                                                  6   
3  7,8,15,17,51,53,58,84,97,113,128,141,154,181,2...   
4                                       11,7572,7911   

                                     indent_item_ids  \
0                  2,3,10,98,99,2459,4772,8274,13100   
1                            4,11,12,2832,2833,13486   
2                                                  5   
3  6,7,8,9,20,21,22,23,24,26,27,88,89,90,92,93,94...   
4                                  13,14,13561,14169   

                                             grn_ids  \
0          10,11,12,17,1119,1243,1467,2841,4786,7447   
1                                    13,18,1788,7679   
2                                                 14   
3  15,16,23,25,78,80,83,113,127,143,158,170,184,2...   
4                                       19,7722,8063   

                                        grn_item_ids  indent_count  grn_count  \
0            1,2,3,10,1821,2025,2420,4908,8345,13178             9         10   
1                            4,11,12,2987,2988,13564             4          4   
2                                                  5             1          1   
3  6,7,8,9,20,21,22,23,24,26,27,74,75,76,77,79,80...            16         16   
4                                  13,14,13638,14251             3          3   

   indent_item_count  grn_item_count  lots_supplied  
0                  9              10              5  
1                  6               6              3  
2                  1               1              0  
3                 52              52              2  
4                  4               4              2

In [21]:
test_df.describe().T

count           mean            std       min  \
farmer                       551.0     291.119782     165.036096    1.0000   
avg_time_to_grn              551.0      99.121575     378.688176    3.3333   
total_expected_qty           551.0  143334.462795  401387.713802  140.0000   
avg_expected_price           551.0     329.605781      77.059775    0.0000   
total_actual_quantity        551.0  141723.716878  400490.202795  118.0000   
avg_actual_price              61.0     761.697493    3493.980183  205.0000   
avg_received_boxes           532.0      89.417331      63.014067    3.0000   
avg_quantity_variance        551.0     539.973850     482.347060    0.0000   
avg_price_variance            61.0     455.398266    3502.654997  -39.0000   
avg_expected_count_variance  551.0      10.187644      36.740285    0.0000   
indent_count                 551.0      14.940109      41.284050    1.0000   
grn_count                    551.0      14.883848      41.185668    1.0000   
indent_item_count            551.0      26.056261      66.625947    1.0000   
grn_item_count               551.0      26.584392      67.992938    1.0000   
lots_supplied                551.0      23.691470      61.895712    0.0000   

                                     25%         50%           75%  \
farmer                        152.500000    294.0000    433.000000   
avg_time_to_grn                54.365150     70.8868     91.973250   
total_expected_qty           7100.000000  23000.0000  86240.000000   
avg_expected_price            284.475000    314.0000    359.056600   
total_actual_quantity        6892.000000  22867.0000  79593.000000   
avg_actual_price              285.000000    305.3333    342.000000   
avg_received_boxes             52.865475     75.0000    104.216825   
avg_quantity_variance         191.833300    427.7500    765.346150   
avg_price_variance             -1.000000      5.0000     10.000000   
avg_expected_count_variance     5.000000      7.1552     10.921850   
indent_count                    1.000000      3.0000      9.000000   
grn_count                       1.000000      3.0000      9.000000   
indent_item_count               2.000000      5.0000     16.500000   
grn_item_count                  2.000000      5.0000     16.500000   
lots_supplied                   1.500000      4.0000     15.000000   

                                      max  
farmer                       5.740000e+02  
avg_time_to_grn              8.832000e+03  
total_expected_qty           3.418625e+06  
avg_expected_price           9.000000e+02  
total_actual_quantity        3.475924e+06  
avg_actual_price             2.760000e+04  
avg_received_boxes           4.920000e+02  
avg_quantity_variance        3.708111e+03  
avg_price_variance           2.736300e+04  
avg_expected_count_variance  8.575000e+02  
indent_count                 4.730000e+02  
grn_count                    4.730000e+02  
indent_item_count            5.820000e+02  
grn_item_count               5.920000e+02  
lots_supplied                5.430000e+02

In [22]:
# Drop avg_actual_price and avg_price_variance due to low volume of data
dataset = test_df.drop(
    columns=[
        "avg_actual_price",
        "avg_price_variance",
        "grn_ids",
        "grn_item_ids",
        "indent_item_ids",
        "indent_ids",
        "lots_supplied",
    ]
)

In [23]:
dataset.describe().T

count           mean            std       min  \
farmer                       551.0     291.119782     165.036096    1.0000   
avg_time_to_grn              551.0      99.121575     378.688176    3.3333   
total_expected_qty           551.0  143334.462795  401387.713802  140.0000   
avg_expected_price           551.0     329.605781      77.059775    0.0000   
total_actual_quantity        551.0  141723.716878  400490.202795  118.0000   
avg_received_boxes           532.0      89.417331      63.014067    3.0000   
avg_quantity_variance        551.0     539.973850     482.347060    0.0000   
avg_expected_count_variance  551.0      10.187644      36.740285    0.0000   
indent_count                 551.0      14.940109      41.284050    1.0000   
grn_count                    551.0      14.883848      41.185668    1.0000   
indent_item_count            551.0      26.056261      66.625947    1.0000   
grn_item_count               551.0      26.584392      67.992938    1.0000   

                                     25%         50%           75%  \
farmer                        152.500000    294.0000    433.000000   
avg_time_to_grn                54.365150     70.8868     91.973250   
total_expected_qty           7100.000000  23000.0000  86240.000000   
avg_expected_price            284.475000    314.0000    359.056600   
total_actual_quantity        6892.000000  22867.0000  79593.000000   
avg_received_boxes             52.865475     75.0000    104.216825   
avg_quantity_variance         191.833300    427.7500    765.346150   
avg_expected_count_variance     5.000000      7.1552     10.921850   
indent_count                    1.000000      3.0000      9.000000   
grn_count                       1.000000      3.0000      9.000000   
indent_item_count               2.000000      5.0000     16.500000   
grn_item_count                  2.000000      5.0000     16.500000   

                                      max  
farmer                       5.740000e+02  
avg_time_to_grn              8.832000e+03  
total_expected_qty           3.418625e+06  
avg_expected_price           9.000000e+02  
total_actual_quantity        3.475924e+06  
avg_received_boxes           4.920000e+02  
avg_quantity_variance        3.708111e+03  
avg_expected_count_variance  8.575000e+02  
indent_count                 4.730000e+02  
grn_count                    4.730000e+02  
indent_item_count            5.820000e+02  
grn_item_count               5.920000e+02

Performing TOPSIS method to rank the farmer performance
1. Normalize the data
2. Calculate the weighted normalized decision matrix
3. Calculate the separation measures
4. Calculate the relative closeness to the ideal solution
5. Rank the farmers based on the relative closeness to the ideal solution



In [24]:
# Normalizing the data and normalizing if we have weights
def normalize_data(dataset, nCol, weights=None):
    for i in range(1, nCol):
        temp = 0

        for j in range(len(dataset)):
            temp += dataset.iloc[j, i] ** 2
        temp = temp**0.5

        for j in range(len(dataset)):
            if weights is not None:
                dataset.iloc[j, i] = (dataset.iloc[j, i] / temp) * weights[i - 1]
            else:
                dataset.iloc[j, i] = dataset.iloc[j, i] / temp

    return dataset


def calculate_ideal_values(dataset, nCol, impact):
    ideal_best = []
    ideal_worst = []

    for i in range(1, nCol):
        if impact[i - 1] == "+":
            ideal_best.append(dataset.iloc[:, i].max())
            ideal_worst.append(dataset.iloc[:, i].min())
        else:
            ideal_best.append(dataset.iloc[:, i].min())
            ideal_worst.append(dataset.iloc[:, i].max())

    return ideal_best, ideal_worst


def calculate_separation_measures(dataset, ideal_best, ideal_worst, nCol):
    separation_best = []
    separation_worst = []

    for i in range(len(dataset)):
        best = 0
        worst = 0

        for j in range(1, nCol):
            best += (dataset.iloc[i, j] - ideal_best[j - 1]) ** 2
            worst += (dataset.iloc[i, j] - ideal_worst[j - 1]) ** 2

        separation_best.append(best**0.5)
        separation_worst.append(worst**0.5)

    return separation_best, separation_worst


def calculate_relative_closeness(separation_best, separation_worst):
    relative_closeness = []

    for i in range(len(separation_best)):
        if separation_best[i] + separation_worst[i] == 0:
            relative_closeness.append(0)
        else:
            relative_closeness.append(
                separation_worst[i] / (separation_best[i] + separation_worst[i])
            )

    return relative_closeness


def rank_farmers(dataset, nCol, impact, weights=None):
    normalized_data = normalize_data(dataset.copy(), nCol, weights)
    ideal_best, ideal_worst = calculate_ideal_values(normalized_data, nCol, impact)
    separation_best, separation_worst = calculate_separation_measures(
        normalized_data, ideal_best, ideal_worst, nCol
    )
    relative_closeness = calculate_relative_closeness(separation_best, separation_worst)
    normalized_data["relative_closeness"] = relative_closeness
    normalized_data["rank"] = normalized_data["relative_closeness"].rank(
        ascending=False
    )

    return normalized_data.sort_values(by="rank")


numeric_columns_impact_mapper = {
    "avg_time_to_grn": "-",
    "total_expected_qty": "+",
    "avg_expected_price": "-",
    "total_actual_quantity": "+",
    "avg_received_boxes": "+",
    "avg_quantity_variance": "-",
    "avg_expected_count_variance": "-",
    "indent_count": "+",
    "grn_count": "+",
    "indent_item_count": "+",
    "grn_item_count": "+",
}

farmer_ids = dataset[["farmer", "farmer_name"]].copy()
numeric_dataset = dataset[
    ["farmer"] + list(numeric_columns_impact_mapper.keys())
].copy()

numeric_dataset = numeric_dataset.fillna(numeric_dataset.mean())

nCol = numeric_dataset.shape[1]

impact = list(numeric_columns_impact_mapper.values())  # Adjusted for 10 numeric columns

equal_weight_factor = 1 / (nCol - 1)  # Equal weight factor for normalization

weights = [equal_weight_factor] * (nCol - 1)

ranked_farmers = rank_farmers(numeric_dataset, nCol, impact, weights)

# Merge with original dataset to get all columns, then add ranking information
final_results = dataset.merge(
    ranked_farmers[["farmer", "relative_closeness", "rank"]], on="farmer", how="left"
)
final_results = final_results.sort_values(by="rank")
final_results.head()

/tmp/ipykernel_1686849/3964253782.py:12: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.000794541531191457' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  dataset.iloc[j, i] = (dataset.iloc[j, i] / temp) * weights[i-1]
/tmp/ipykernel_1686849/3964253782.py:12: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.0008850749635979672' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  dataset.iloc[j, i] = (dataset.iloc[j, i] / temp) * weights[i-1]
/tmp/ipykernel_1686849/3964253782.py:12: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.0004876051448425643' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  dataset.iloc[j, i] = (datas

farmer     farmer_name  avg_time_to_grn  total_expected_qty  \
17       21         KS RAJU          74.3528           3391850.0   
154     169   RAMALINGARAJU         200.7969           3106274.0   
15       19  SRINIVASARAO K          77.4832           3418625.0   
14       18         K SRINU         107.8041           2036650.0   
36       40  SRI RAMA AQUA           81.9508           2539562.0   

     avg_expected_price  total_actual_quantity  avg_received_boxes  \
17             289.8090              3463088.0            133.9238   
154            306.6151              3125085.0             97.1199   
15             312.4812              3475924.0             94.8579   
14             309.8884              1922421.0             77.6538   
36             310.8431              2517238.0             70.5150   

     avg_quantity_variance  avg_expected_count_variance  indent_count  \
17               1169.1873                      11.3017           473   
154               648.4749                       8.5735           400   
15               1732.8690                      17.1245           361   
14               1399.0011                      18.2130           239   
36                678.9086                      10.8840           178   

     grn_count  indent_item_count  grn_item_count  relative_closeness  rank  
17         473                563             576            0.910612   1.0  
154        400                556             556            0.887891   2.0  
15         360                582             592            0.866155   3.0  
14         239                386             406            0.760932   4.0  
36         177                341             343            0.742269   5.0

In [25]:
final_results.head(50)

farmer                         farmer_name  avg_time_to_grn  \
17       21                             KS RAJU          74.3528   
154     169                       RAMALINGARAJU         200.7969   
15       19                      SRINIVASARAO K          77.4832   
14       18                             K SRINU         107.8041   
36       40                      SRI RAMA AQUA           81.9508   
388     407     SHREEYAN MARINE EXPORTS PVT LTD          53.4301   
49       53                        NAGESH VARMA          80.6364   
265     284                          MASTAN SKM          62.6291   
35       39                  CHANDRIKA SEAFOODS          71.1003   
73       77                   PK RAMALINGA RAJU          77.2637   
66       70                        B PANDURANGA          83.1048   
99      105                     SYAMALA TRADERS          81.2544   
253     272                             K RAGHU         105.5844   
220     237  SANDHYA AQUA FARMS KRITHIVENNU OWN          63.2885   
484     507                  VELANKKANI TRADERS          48.2457   
9        13                 MEENAKSHI FISHEREIS         108.3900   
28       32                RAVINDRA ENTERPRISES          61.0184   
89       94                                SKTC          84.2936   
78       82                             SV AQUA          72.5674   
147     162                                KVSN          76.3183   
44       48                     PVR ENTERPRISES          98.4272   
31       35                          SS TRADERS          71.3807   
250     268                     ARUNODAYA AQUA           67.7689   
18       22                        SKM SEAFOODS         125.8602   
305     324                       ANJANEYA AQUA          68.6568   
45       49                            R KARUNA          90.9378   
72       76                       SIVA SEAFOODS          82.7271   
6        10                      Sita rama raju          77.9415   
390     409     SREEYAAN MARINE EXPORTS PVT LTD          58.3802   
309     328                          NVVS VARMA         120.3828   
11       15               VENKATESWARA SEAFOODS          78.4536   
433     454                       PV TRADELINKS          49.8689   
34       38                     VENKANNA BABU Y          68.2727   
256     275                                SRSF          78.7123   
155     170                      M NAVEEN REDDY          60.8792   
126     141                      AQUA EXCHANGE           87.4361   
129     144                     P RAMALINGARAJU          85.5563   
176     191                                 SRK          68.0000   
193     208            P LAKSHMI NARAYANA VARMA          71.1786   
57       61                   NAGAMALLESWARAO K          69.1624   
207     222                        GANESH VARMA          76.6875   
537     560                   LAVANYA SEA FOODS          48.9488   
201     216                     NN SRINIVASARAO         120.4777   
431     452                          NN TRADERS          54.4776   
113     127                             RAMBABU          56.4326   
401     420                ADITYA AQUA AGENCIES          57.4219   
468     490                    GUDURU OWN PONDS          56.0876   
16       20             SAI LAKSHMI AUQA FARMS           99.0094   
22       26                       SWAMY NAIDU K         110.7243   
39       43               VENKATESWARA SEAFOODS          84.1803   

     total_expected_qty  avg_expected_price  total_actual_quantity  \
17            3391850.0            289.8090              3463088.0   
154           3106274.0            306.6151              3125085.0   
15            3418625.0            312.4812              3475924.0   
14            2036650.0            309.8884              1922421.0   
36            2539562.0            310.8431              2517238.0   
388           2701530.0            304.5837              2606042.0   
49            1872480.0            298.5796   

In [30]:
# Save to Excel file
final_results.to_excel("farmer_performance_ranking.xlsx", index=False)